userspace/plugin/plugin_loader.c

```diff
@@ -0,0 +1,235 @@
+    }
+
+    // load all library symbols
+    SYM_RESOLVE(ret, get_required_api_version);
+    SYM_RESOLVE(ret, get_version);
+    SYM_RESOLVE(ret, get_last_error);
+    SYM_RESOLVE(ret, get_name);
+    SYM_RESOLVE(ret, get_description);
+    SYM_RESOLVE(ret, get_contact);
+    SYM_RESOLVE(ret, get_init_schema);
+    SYM_RESOLVE(ret, init);
+    SYM_RESOLVE(ret, destroy);
+    SYM_RESOLVE(ret, get_id);
+    SYM_RESOLVE(ret, get_event_source);
+    SYM_RESOLVE(ret, open);
+    SYM_RESOLVE(ret, close);
+    SYM_RESOLVE(ret, next_batch);
+    SYM_RESOLVE(ret, get_progress);
+    SYM_RESOLVE(ret, list_open_params);
+    SYM_RESOLVE(ret, event_to_string);
+    SYM_RESOLVE(ret, get_fields);
+    SYM_RESOLVE(ret, extract_fields);
+    SYM_RESOLVE(ret, get_extract_event_sources);
+    return ret;
+}
+
+void plugin_unload(plugin_handle_t* h)
+{
+    if (h)
+    {
+        if (h->handle)
+        {
+#ifdef _WIN32
+            FreeLibrary(h->handle);
+#else
+            dlclose(h->handle);
+#endif
+        }
+        free(h);
+    }
+}
+
+bool plugin_is_loaded(const char* path)
+{
+#ifdef _WIN32
+	/*
+	 * LoadLibrary maps the module into the address space of the calling process, if necessary,
+	 * and increments the modules reference count, if it is already mapped.
+	 * GetModuleHandle, however, returns the handle to a mapped module
+	 * without incrementing its reference count.
+	 *
+	 * This returns an HMODULE indeed, but they are the same thing
+	 */
+	return GetModuleHandle(path) != NULL;
+#else
+	/*
+	 * RTLD_NOLOAD (since glibc 2.2)
+	 *	Don't load the shared object. This can be used to test if
+	 *	the object is already resident (dlopen() returns NULL if
+	 *	it is not, or the object's handle if it is resident).
+	 *	This does not increment dlobject reference count.
+	 */
+	return dlopen(path, RTLD_LAZY | RTLD_NOLOAD) != NULL;
+#endif
+}
+
+bool plugin_check_required_api_version(const plugin_handle_t* h, char* err)
+{
+    uint32_t major, minor, patch;
+    const char *ver, *failmsg;
+    if (h->api.get_required_api_version == NULL)
+    {
+        strncpy(err, "plugin_get_required_api_version symbol not implemented", PLUGIN_MAX_ERRLEN - 1);
+        return false;
+    }
+
+    ver = h->api.get_required_api_version();
+    if (sscanf(ver, "%" PRIu32 ".%" PRIu32 ".%" PRIu32, &major, &minor, &patch) != 3)
+    {
+        snprintf(err, PLUGIN_MAX_ERRLEN - 1, "plugin provided an invalid required API version: '%s'", ver);
+        return false;
+    }
+
+    failmsg = NULL;
+    if(PLUGIN_API_VERSION_MAJOR != major)
+    {
+        failmsg = "major versions disagree";
+    }
+    else if(PLUGIN_API_VERSION_MINOR < minor)
+    {
+        failmsg = "framework's minor is less than the requested one";
+    }
+    else if(PLUGIN_API_VERSION_MINOR == minor && PLUGIN_API_VERSION_PATCH < patch)
+    {
+        failmsg = "framework's patch is less than the requested one";
+    }
+
+    if (failmsg != NULL)
+    {
+        snprintf(err, PLUGIN_MAX_ERRLEN - 1,
```

Comment:
On last ultra-nit, the size in `snprintf` includes the terminating null byte, so you don't need to `- 1` it.
```suggestion
        snprintf(err, PLUGIN_MAX_ERRLEN,
```

Reply:
Oh thanks! I forgot about that!